language/move-prover/interpreter/src/concrete/player.rs

```diff
@@ -0,0 +1,1357 @@
+    let addr = op_addr.deref().into_address();
+    match global_state.del_resource(addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_get_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_addr: RefTypedValue,
+    global_state: &mut GlobalState,
+) -> Result<TypedValue, AbortInfo> {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    match global_state.get_resource(None, addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_borrow_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    is_mut: bool,
+    op_addr: RefTypedValue,
+    global_state: &mut GlobalState,
+) -> Result<TypedValue, AbortInfo> {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    match global_state.get_resource(Some(is_mut), addr, inst) {
+        None => Err(AbortInfo::sys_abort(StatusCode::RESOURCE_DOES_NOT_EXIST)),
+        Some(object) => Ok(object),
+    }
+}
+
+fn handle_exists_global(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_addr: RefTypedValue,
+    global_state: &GlobalState,
+) -> TypedValue {
+    let env = ctxt.target.global_env();
+    let inst = convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+    let addr = op_addr.deref().into_address();
+    TypedValue::mk_bool(global_state.has_resource(&addr, &inst))
+}
+
+fn handle_write_back_global_struct(
+    ctxt: &FunctionContext,
+    module_id: ModuleId,
+    struct_id: StructId,
+    ty_args: &[MT::Type],
+    op_struct: RefTypedValue,
+    global_state: &mut GlobalState,
+) {
+    let (struct_ty, object, ptr) = op_struct.deref().decompose();
+    let inst = struct_ty.into_ref_struct_inst(Some(true));
+    if cfg!(debug_assertions) {
+        let env = ctxt.target.global_env();
+        let converted =
+            convert_model_struct_type(env, module_id, struct_id, ty_args, &ctxt.ty_args);
+        assert_eq!(inst, converted);
+    }
+    let addr = match ptr {
+        Pointer::Global(addr) => addr,
+        _ => unreachable!(),
+    };
+    global_state.put_resource(addr, inst, object);
+}
+
+fn handle_write_back_local(
+    local_root: TempIndex,
+    op_val: TypedValue,
+    local_state: &mut LocalState,
+) {
+    let (ty, val, ptr) = op_val.decompose();
+    if cfg!(debug_assertions) {
+        assert!(matches!(ptr, Pointer::Local(root_idx) if root_idx == local_root));
```

Comment:
This should not be an assert. Rather it must be a test whether root_idx == local_root and only if it succeeds we perform the update.

Consider

```
let a = 1; let b = 2;
let r = if (b) { &mut a } else  {&mut b };
*r = 3;
```

We will create two WriteBacks:

```
write_back[Local(a)](r)
write_back[Local(b)](r)
```
Only one of them will be effective, depending on what we find Pointer::Local.

A similar logic is currently missing for global write-backs in the borrow graph and therefore in the generated bytecode, therefore the TODO I mentioned above.

Reply:
Ah, I see. Thanks for the clarification! Will update.